### PR TITLE
Add safe module

### PR DESCRIPTION
**Pull request by JohnGuilding**
This PR:
- Adds a hardhat project
- Adds a safe plugin with EOA sig verification.
- Adds a test to run a user operation through the plugin via a bundler (this required a hardhat project, hence the webauthn directory now contains a joint foundry and hardhat project.

These changes were based off of Safe's own code (module and test linked below). To get a better understanding of the changes I originally made to the Safe files, see these changes I made in [this branch](https://github.com/JohnGuilding/safe-contracts/tree/update-4337-plugin-and-run-geth-node) (from a safe-contracts fork). To then review the changes made to that code, see commit: `Convert integration test to ethers v6`.

## Safe plugin
The Safe plugin was based off of [Safe's POC 4337 module and handler](https://github.com/safe-global/safe-contracts/blob/main/contracts/test/4337/Test4337ModuleAndHandler.sol).

Main changes I made:
1. I changed the contract to inherit from a `BaseAccount`, so we had the correct interface for `_validateSignature`.
2. Added an owner state variable
3. Added EOA validation as the contract did not run any signature validation before

## Integration test
The integration test was based off of [Safe's accompanying test](https://github.com/safe-global/safe-contracts/blob/main/test/integration/Safe.ERC4337.spec.ts) for their module and handler

Main changes I made:
1. Used `eth_sendUserOperation` instead of `eth_estimateUserOperationGas`
2. Added EOA signature to user op
3. Added an assertion to verify the user op transfered ETH to a recipient
4. Used ethers v6 (except for the `bunderProvider` (see comment on that))

# How to manually test these changes:
I added instructions to run the integration test in the webauthn readme. We don't have CI set up yet so these are manual tests. Additionally, you'll need to copy some addresses to a `.env` file as the deployments addresses are not deterministic yet.

### CODE REVIEW

File: webauthn/test/integration/SafeWebAuthn.test.ts [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+const MNEMONIC = process.env.MNEMONIC;
+
+const sleep = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));
+
+describe("Safe.ERC4337", () => {
+  const setupTests = async () => {
+    const factory = await hre.ethers.getContractFactory("SafeProxyFactory");
+    const singleton = await hre.ethers.getContractFactory("Safe");
+
+    const bundlerProvider = new ethersV5.providers.JsonRpcProvider(BUNDLER_URL);
```
Comment (JohnGuilding) [inline webauthn/test/integration/SafeWebAuthn.test.ts:31]: The `bundlerProvider` did not play nicely with ethers v6. So installed ethers v5 and aliased it (see ethers-v5 in package.json).

Example of the errors I was getting with an ethers v6 provider attached to the bundler:
```
JsonRpcProvider failed to detect network and cannot start up; retry in 1s (perhaps the URL is wrong or the node is not started)
```
```
    Error: missing response for request (value=[ { "error": { "code": -32601,
    "message": "Method undefined is not supported" } } ], info={ "payload": { "id": 2,
    "jsonrpc": "2.0", "method": "eth_supportedEntryPoints", "params": [  ] } },
    code=BAD_DATA, version=6.6.7)
```

File: webauthn/src/utils/SafeTypechainTypesHelper.sol
Comment (JohnGuilding) [inline webauthn/src/utils/SafeTypechainTypesHelper.sol:1]: I did try and find the typechain types in the npm packages that Safe provide, however, I couldn't find them. This was my work around to get the required typechain types.

File: webauthn/src/SafeWebAuthnPlugin.sol [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    function owner() public view returns (address) {
+        return _owner;
+    }
+
+    function _validateSignature(
+        UserOperation calldata userOp,
+        bytes32 userOpHash
+    ) internal override returns (uint256 validationData) {
+        bytes32 hash = userOpHash.toEthSignedMessageHash();
+        if (_owner != hash.recover(userOp.signature))
```
Comment (blakecduncan) [inline webauthn/src/SafeWebAuthnPlugin.sol:87]: Is the idea that this plugin is currently validating sigs with ECDSA signatures but you will eventually turn this into webauthn sigs?  Do you think there's any benefit to keeping this as an EOA pluggin and then just creating the webAuthn one separately?

**Review by blakecduncan — APPROVED** [review]
Nice this looks good!  I think the only thought I had was I wonder if it's worth while just creating an ECDSA plugin and test.  I'm guessing your plan is to eventually swap this out for webauthn validation?

File: webauthn/test/integration/SafeWebAuthn.test.ts [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+      [recipientAddress, transferAmount, "0x00"]
+    );
+
+    // Native tokens for the pre-fund 💸
+    await userWallet.sendTransaction({
+      to: deployedAddress,
+      value: ethers.parseEther("100"),
+    });
+    // The bundler uses a different node, so we need to allow it sometime to sync
+    await sleep(5000);
```
Comment (jzaki) [inline webauthn/test/integration/SafeWebAuthn.test.ts:155]: Can find a way around these later.

**Review by jzaki — APPROVED** [review]

File: webauthn/src/SafeWebAuthnPlugin.sol (hunk 2 above)
Comment (JohnGuilding) [reply webauthn/src/SafeWebAuthnPlugin.sol:87]: > Is the idea that this plugin is currently validating sigs with ECDSA signatures but you will eventually turn this into webauthn sigs?

Yeah the original plan was to swap this out for webauthn sigs. 

> Do you think there's any benefit to keeping this as an EOA pluggin and then just creating the webAuthn one separately?

I don't have a strong opinion. For my work rn, it's probably doesn't add much benefit to have the ECSDA sig plugin too.  I know James is looking into composable verification logic. In that case, it might be useful for testing and building that. I can leave it separate for now and then it's there if we need it :)

File: webauthn/src/SafeWebAuthnPlugin.sol [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        uint256 value,
+        bytes memory data,
+        uint8 operation
+    ) external returns (bool success);
+}
+
+contract SafeWebAuthnPlugin is BaseAccount {
+    using ECDSA for bytes32;
+
+    address public immutable myAddress;
```
[the remark this answers is not in the document]
Comment (JohnGuilding) [reply webauthn/src/SafeWebAuthnPlugin.sol:25]: So this module is specific to a single safe wallet, and some of the others I've seen look they do the same thing. Although it is possible to also do things the Kernel way:

For example, [this allowance module](https://github.com/safe-global/safe-modules/blob/master/allowances/contracts/AlowanceModule.sol) only needs to be deployed once per network. From the allowance module readme:
> The contract is designed as a single point registry. This way not every Safe needs to deploy their own module and it is possible that this module is shared between different Safes

My understanding isn't too deep, so I'm not sure if there are certain restrictions to what you can do when choosing between a singleton, vs being deployed once for each wallet.

File: webauthn/src/SafeWebAuthnPlugin.sol [hunk 5]
```diff
@@ -0,0 +1,91 @@
+// SPDX-License-Identifier: LGPL-3.0-only
+pragma solidity >=0.7.0 <0.9.0;
+pragma abicoder v2;
+
+import {BaseAccount} from "account-abstraction/contracts/core/BaseAccount.sol";
+import {IEntryPoint, UserOperation} from "account-abstraction/contracts/interfaces/IEntryPoint.sol";
+import {UserOperation} from "account-abstraction/contracts/interfaces/IEntryPoint.sol";
```
[the remark this answers is not in the document]
Comment (JohnGuilding) [reply webauthn/src/SafeWebAuthnPlugin.sol:7]: Thanks, will update

File: webauthn/src/SafeWebAuthnPlugin.sol [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+    ) external payable {
+        address payable safeAddress = payable(msg.sender);
+        ISafe safe = ISafe(safeAddress);
+        require(
+            safe.execTransactionFromModule(to, value, data, 0),
+            "tx failed"
+        );
+    }
+
+    function enableMyself() public {
```
[the remark this answers is not in the document]
Comment (JohnGuilding) [reply webauthn/src/SafeWebAuthnPlugin.sol:70]: Good shout, will add a follow up issue

File: webauthn/test/integration/SafeWebAuthn.test.ts (hunk 3 above)
Comment (JohnGuilding) [reply webauthn/test/integration/SafeWebAuthn.test.ts:155]: Yeah that would be nice

File: webauthn/script/start.sh [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+docker pull ethereum/client-go:v1.10.26
+
+CONTAINER=geth$RANDOM
+
+docker run --rm -i --name $CONTAINER -p 8545:8545 ethereum/client-go:v1.10.26 \
+  --miner.gaslimit 100000000 \
+  --http \
+  --http.api personal,eth,net,web3,debug \
+  --http.vhosts '*,localhost,host.docker.internal' \
+  --http.corsdomain='*' \
```
[the remark this answers is not in the document]
Comment (JohnGuilding) [reply webauthn/script/start.sh:24]: Will do this in follow up PR

File: webauthn/script/fundAccounts.js [hunk 8]
```diff
@@ -0,0 +1,11 @@
+const devAccounts = [
```
[the remark this answers is not in the document]
Comment (JohnGuilding) [reply webauthn/script/fundAccounts.js:1]: This file is copied [from bls-wallet](https://github.com/getwax/bls-wallet/blob/main/contracts/scripts/fundAccounts.js) and was how we funded accounts on geth there. They are [the first 5 accounts hardhat uses](https://hardhat.org/hardhat-network/docs/reference#initial-state) so they should be in sync. Let me check if I can get it from a provider or similar, if I can, I'll use that

File: webauthn/src/SafeWebAuthnPlugin.sol (hunk 2 above)
Comment (JohnGuilding) [reply webauthn/src/SafeWebAuthnPlugin.sol:87]: >  another thought is we could rename this project in the mono repo to be safe-plugins rather than webauthn

Actually I don't think this has been posted in the chat yet but something James and I discussed last week was monorepo structure. I think the rough plan could be:
- Use the webauth directory now as the integrations directory, as it already has foundry and hardhat setup.
    - The integrations directory would then have directories for Safe, Kernel and any other project we integrated with.
    - We'd build our higher level modules here. So both our current work would live in this 'integrations' directory
- Then we'd have a plain foundry project for the low level cryptographic logic e.g. bls and webauthn - that could be used by the high level kernel and safe modules

**TLDR:** similar to what you outlined but have the kernel and safe plugins in one directory (hardhat & foundry), and then have the low level cryptography stuff in a single dedicated directory (just foundry). What do you think to that idea?

I think James is outlining this now under his standup note titled "proposed github structure".

Comment (JohnGuilding) [reply webauthn/src/SafeWebAuthnPlugin.sol:87]: I guess for the ecsda plugin, under the potential new structure, that logic would live in the low level crypto dir, and then could be used by the kernel and safe plugins in the integrations dir

File: webauthn/src/SafeWebAuthnPlugin.sol [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+
+    address internal constant _SENTINEL_MODULES = address(0x1);
+
+    constructor(address entryPointAddress, address ownerAddress) {
+        myAddress = address(this);
+        _owner = ownerAddress;
+        _entryPoint = entryPointAddress;
+    }
+
+    function validateUserOp(
```
[the remark this answers is not in the document]
Comment (JohnGuilding) [reply webauthn/src/SafeWebAuthnPlugin.sol:37]: #46

File: webauthn/src/SafeWebAuthnPlugin.sol [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+        if (missingAccountFunds != 0) {
+            senderSafe.execTransactionFromModule(
+                _entryPoint,
+                missingAccountFunds,
+                "",
+                0
+            );
+        }
+
+        validationData = _validateSignature(userOp, userOpHash);
```
[the remark this answers is not in the document]
Comment (JohnGuilding) [reply webauthn/src/SafeWebAuthnPlugin.sol:54]: #46

File: webauthn/src/SafeWebAuthnPlugin.sol (hunk 6 above)
[the remark this answers is not in the document]
Comment (JohnGuilding) [reply webauthn/src/SafeWebAuthnPlugin.sol:70]: #47
